MSVC linkage errors

I'm getting linkage errors in C++ code that uses process.h; the predeclarations of various internal Windows API calls doesn't match what's eventually defined by windows.h, if both headers are included:
> 1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\handleapi.h(39,6): error C2732: linkage specification contradicts earlier specification for 'CloseHandle'
> 1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\handleapi.h(38): message : see declaration of 'CloseHandle'
> 1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\handleapi.h(87,6): error C2732: linkage specification contradicts earlier specification for 'SetHandleInformation'
> 1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\handleapi.h(84): message : see declaration of 'SetHandleInformation'
> 1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\namedpipeapi.h(34,6): error C2732: linkage specification contradicts earlier specification for 'CreatePipe'
> 1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\namedpipeapi.h(30): message : see declaration of 'CreatePipe'
> 1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\synchapi.h(382,6): error C2732: linkage specification contradicts earlier specification for 'WaitForSingleObject'
> 1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\synchapi.h(380): message : see declaration of 'WaitForSingleObject'
> 1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\processthreadsapi.h(154,6): error C2732: linkage specification contradicts earlier specification for 'GetExitCodeProcess'
> 1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\processthreadsapi.h(152): message : see declaration of 'GetExitCodeProcess'
> 1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\processthreadsapi.h(371,6): error C2732: linkage specification contradicts earlier specification for 'CreateProcessA'
> 1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\processthreadsapi.h(361): message : see declaration of 'CreateProcessA'

You can reproduce the problem by adding `#include <windows.h>` at the end of test.cpp.

The error starts occurring in 5bdac3a7eade787; I _think_ it has something to do with the narrowing of the `extern "C"` scope, which no longer includes the predeclarations. They're always defined as `extern "C"` in windows.h; moving them inside an `extern "C"` scope fixes the linkage error. Is there a reason they were moved outside that scope?

I got the same thing. If include Windows.h before process.h, no problem, but if I don't include it, I get link errors.

eg. 
```
error LNK2019: unresolved external symbol "__declspec(dllimport) int __cdecl SetHandleInformation(void *,unsigned long,unsigned long)" (__imp_?SetHandleInformation@@YAHPEAXKK@Z) referenced in function subprocess_create
```

Which is weird, since process.h declares them as __stdcall, not __cdecl 🤔

And if I include Windows.h after, I get the same errors as you.

--- a/.travis.yml
+++ b/.travis.yml
@@ -19,4 +19,4 @@ script:
   - cd build
   - cmake -DCMAKE_BUILD_TYPE=$CONFIGURATION ../test
   - make
-  - ./process_test
+  - ./subprocess_test

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -36,4 +36,4 @@ build_script:
   - if "%PLATFORM%"=="Win32" cmake -G "%VSVERSION%" ../test
   - msbuild /m /p:Configuration="%CONFIGURATION%" /p:Platform="%PLATFORM%" process.sln
   - cd %CONFIGURATION%
-  - process_test.exe
+  - subprocess_test.exe

--- a/subprocess.h
+++ b/subprocess.h
@@ -125,10 +125,6 @@ subprocess_weak int subprocess_join(struct subprocess_s *const process,
 /// the parent process.
 subprocess_weak int subprocess_destroy(struct subprocess_s *const process);
 
-#if defined(__cplusplus)
-} // extern "C"
-#endif
-
 #if !defined(_MSC_VER)
 #include <stdlib.h>
 #include <sys/types.h>
@@ -570,5 +566,9 @@ int subprocess_destroy(struct subprocess_s *const process) {
 
   return 0;
 }
+
+#if defined(__cplusplus)
+} // extern "C"
+#endif
 
 #endif /* SHEREDOM_SUBPROCESS_H_INCLUDED */

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -41,12 +41,14 @@ add_executable(process_stderr_argv process_stderr_argv.c)
 add_executable(process_combined_stdout_stderr process_combined_stdout_stderr.c)
 add_executable(process_inherit_environment process_inherit_environment.c)
 
-add_executable(process_test
+add_executable(subprocess_test
   ../subprocess.h
   utest.h
   main.c
   test.c
   test.cpp
+  pre_windows.cpp
+  post_windows.cpp
 )
 
 if("${CMAKE_C_COMPILER_ID}" STREQUAL "GNU")

--- a/test/post_windows.cpp
+++ b/test/post_windows.cpp
@@ -1,0 +1,31 @@
+/*
+   This is free and unencumbered software released into the public domain.
+
+   Anyone is free to copy, modify, publish, use, compile, sell, or
+   distribute this software, either in source code form or as a compiled
+   binary, for any purpose, commercial or non-commercial, and by any
+   means.
+
+   In jurisdictions that recognize copyright laws, the author or authors
+   of this software dedicate any and all copyright interest in the
+   software to the public domain. We make this dedication for the benefit
+   of the public at large and to the detriment of our heirs and
+   successors. We intend this dedication to be an overt act of
+   relinquishment in perpetuity of all present and future rights to this
+   software under copyright law.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+   IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+   OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+   ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+   OTHER DEALINGS IN THE SOFTWARE.
+
+   For more information, please refer to <http://unlicense.org/>
+*/
+#include "subprocess.h"
+
+#if defined(_MSC_VER)
+#include <windows.h>
+#endif

--- a/test/pre_windows.cpp
+++ b/test/pre_windows.cpp
@@ -1,0 +1,32 @@
+/*
+   This is free and unencumbered software released into the public domain.
+
+   Anyone is free to copy, modify, publish, use, compile, sell, or
+   distribute this software, either in source code form or as a compiled
+   binary, for any purpose, commercial or non-commercial, and by any
+   means.
+
+   In jurisdictions that recognize copyright laws, the author or authors
+   of this software dedicate any and all copyright interest in the
+   software to the public domain. We make this dedication for the benefit
+   of the public at large and to the detriment of our heirs and
+   successors. We intend this dedication to be an overt act of
+   relinquishment in perpetuity of all present and future rights to this
+   software under copyright law.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+   IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+   OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+   ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+   OTHER DEALINGS IN THE SOFTWARE.
+
+   For more information, please refer to <http://unlicense.org/>
+*/
+
+#if defined(_MSC_VER)
+#include <windows.h>
+#endif
+
+#include "subprocess.h"

